### ISSUE TITLE
Implement a timeout after SIGTERM and SIGINT are received

### ISSUE THREAD

**Issue by palourde** [sensu/sensu-go#2373]
When etcd loses its quorum, which can happen during startup when other cluster members are not started or if members are lost, sensu-backend becomes unresponsive.

This causes a UX issue in my opinion because you only see etcd throwing errors and sensu-backend doesn't respond to signals. This is relative "fine" during etcd startup because it shutdowns itself after 60 seconds, however once etcd is properly started and it loses its quorum, based on the experiences I had, the process will never stop and you have no other options but to force kill the sensu-backend process.

We need to decide what to do in this situation and the potential impact. Some potential solutions:

1. Show a log statement every time a signal is received that tells the user that we are currently attempting to gracefully shutdown etcd
	- That doesn't solve the issue when etcd has quorum and lost it
2. Kill ourselves etcd after a given amount of time
	- We need to consider the impact on the state of the cluster itself, as [mentioned by Greg](https://github.com/sensu/sensu-go/issues/2373#issuecomment-440707370).

**Comment by grepory** [comment]
Signal handlers should be started _before_ backend initialization or as early as possible.

Nothing else should be done for now. We can figure out what to do if etcd isn't available later, but our operating conditions so far have been defined as "if etcd doesn't work, then sensu doesn't work."

If we want to change those operating conditions, then yes--the solution is _very_ complicated.

**Comment by grepory** [comment]
If you don't give the cluster time to decide if it can establish connectivity, then you are dangerously killing the etcd processes and you can get into a state where your cluster will never re-establish itself.

I assume that if the cluster _can_ establish connectivity, then the backend startup resumes normally?

**Comment by echlebek** [comment]
SIGTERM and SIGINT are directed to a channel that is immediate consumed by a goroutine, so I think it's unlikely that we are not trapping the signal. If the backend becames unresponsive before the signals are handled, then the default signal handler would kill the process.

I think that what's happening is backend.Stop() is blocking, which happens when a signal is caught. The Stop() method cancels a context and then waits for all the daemons to shut down including etcd.

**Comment by palourde** [comment]
For reference, it looks like this also happen when I manually start a 3-nodes local cluster and I stop one or two nodes. The other(s) remaining become unresponsive.

**Comment by grepory** [comment]
I would rephrase this ticket. This isn't a bug.

We should decide what happens when etcd doesn't have quorum.

**Comment by palourde** [comment]
That makes total sense, I'll rename it and update the description

**Comment by echlebek** [comment]
I've changed the title of this issue after some discussion with Greg.

I believe that the behaviour we have now is correct, but that we should let users kill the backend more quickly when desired. To this end, we can implement a timeout in our signal handler that issues log.Fatal, forcibly killing etcd if it is still trying to achieve quorum.

In other words, I believe the issue as originally described is valid, from a UX point of view.

**Comment by palourde** [comment]
> we should let users kill the backend more quickly when desired

If the cluster had quorum and loses it, the "remaining" nodes will become completely unresponsive unless you force kill it!
